lib/realtime/tenants/authorization.ex

```diff
@@ -73,37 +92,69 @@ defmodule Realtime.Tenants.Authorization do
     sub = Map.get(claims, :sub)
     claims = Jason.encode!(claims)
     headers = headers |> Map.new() |> Jason.encode!()
+    channel_name = if channel, do: channel.name, else: nil
 
-    Postgrex.transaction(conn, fn conn ->
-      Postgrex.query(
-        conn,
-        """
-        SELECT
-         set_config('role', $1, true),
-         set_config('realtime.channel_name', $2, true),
-         set_config('request.jwt.claim.role', $3, true),
-         set_config('request.jwt', $4, true),
-         set_config('request.jwt.claim.sub', $5, true),
-         set_config('request.jwt.claims', $6, true),
-         set_config('request.headers', $7, true)
-        """,
-        [role, channel_name, role, jwt, sub, claims, headers]
-      )
-
-      case Postgrex.query(conn, "SELECT name from realtime.channels", [], mode: :savepoint) do
-        {:ok, %{num_rows: 0}} ->
-          {:ok, %Permissions{read: false}}
-
-        {:ok, _} ->
-          {:ok, %Permissions{read: true}}
-
-        {:error, %Postgrex.Error{postgres: %{code: :insufficient_privilege}}} ->
-          {:ok, %Permissions{read: false}}
-
-        {:error, error} ->
-          Logger.error("Error getting permissions for connection: #{inspect(error)}")
-          {:error, error}
-      end
-    end)
+    Postgrex.query(
+      conn,
+      """
+      SELECT
+       set_config('role', $1, true),
+       set_config('realtime.channel_name', $2, true),
+       set_config('request.jwt.claim.role', $3, true),
+       set_config('request.jwt', $4, true),
+       set_config('request.jwt.claim.sub', $5, true),
+       set_config('request.jwt.claims', $6, true),
+       set_config('request.headers', $7, true)
+      """,
+      [role, channel_name, role, jwt, sub, claims, headers]
+    )
+  end
+
+  defp check_read_permissions(conn, permissions) do
+    query = from(c in Channel, select: c.name)
+
+    case Repo.all(conn, query, Channel, mode: :savepoint) do
+      {:ok, channels} when channels != [] ->
+        {:ok, %Permissions{permissions | read: true}}
+
+      {:ok, _} ->
+        {:ok, %Permissions{permissions | read: false}}
+
+      {:error, %Postgrex.Error{postgres: %{code: :insufficient_privilege}}} ->
+        {:ok, %Permissions{permissions | read: false}}
+
+      {:error, error} ->
+        Logger.error("Error getting permissions for connection: #{inspect(error)}")
+        {:error, error}
+    end
+  end
+
+  defp check_write_permissions(_, permissions, %__MODULE__{channel: nil}) do
+    {:ok, %Permissions{permissions | write: false}}
+  end
+
+  defp check_write_permissions(conn, permissions, %__MODULE__{channel: channel}) do
+    check = Ecto.UUID.generate()
+    changeset = Channel.check_changeset(channel, %{check: check})
+
+    case Repo.update(conn, changeset, Channel, mode: :savepoint) do
```

Comment:
Could you instead do a `Repo.update_all from(Channel, where: [id: ...]), set: [check: true]`? Even if the value is always the same, it would check for permissions, no? Or is there a reason to use a random UUID instead?

Reply:
Yes it would and to be fair there's no good reason to use UUID 🤔 

will change to use that approach to simplify it

Reply:
can't use Repo.update_all in this scenario to use the right connection which was dynamically being used but still will avoid string and uuids to make it simpler